### ISSUE TITLE
Learned pose classifier as an alternative to the bands

### ISSUE THREAD

**Issue by changtimwu** [changtimwu/airpod-pose-estimate#5]
## Summary

There are two ways to decide whether someone is in a pose. We ship one of them. This issue plans the other, **as an alternative that competes with it on evidence — not as a successor**.

Both live behind the same one-line seam, so they can coexist indefinitely and be swapped with a flag:

```python
def update(self, pitch: float, roll: float, target_id: str) -> Tuple[bool, Optional[str], float]:
    """Returns (match, detected_pose_id, error_deg) for this sample."""
```

`HoldTimer`, the hysteresis, the SSE frame and the whole frontend sit downstream and don't care which one produced `match`. Whatever we learn, **bands stay in the repo permanently** — a model that needs a working GPU driver at 9am on demo day is not a fallback.

---

## The two approaches, honestly

| | **Bands** (shipped) | **Learned classifier** (proposed) |
|---|---|---|
| Data needed | None | ~150 labelled holds, ≥3 people |
| Detection latency | ~1 sample (20 ms) | One window (~1.5 s) |
| Handles body/range variation | Poorly — one rectangle for everyone | This is the whole point |
| Handles a new pose | Write 4 numbers | Re-record, re-train |
| Debuggable at 2am | Read the JSON | Read a confusion matrix |
| Fails by | Being wrong for people unlike the tuner | Being confidently wrong off-distribution |
| Cost to change | One line | One recording session |

Neither is strictly better. Bands are excellent when poses are well separated on one or two axes and someone has actually measured them. A classifier earns its place when poses overlap in the angles, when individual variation dominates, or when we want a *quality* score rather than a boolean.

### The thing to be clear-eyed about

The bands have not lost a fair fight yet. Their two real failures were a **placement mismatch** (bands written for a bud on a forearm, buds worn in the ears) and a **convention mismatch** (a level head reads pitch −90, not 0, so a band centred on 0 could never match). Every band in `poses.json` still says so in its own header — *"EVERY BAND BELOW IS A GUESS"*.

A model trained on the same zero data would have failed identically. So the comparison only becomes meaningful **after** someone measures what the poses actually look like — and that measurement is what makes the bands good too. Which is the useful part: **the data work below is not a bet on the model.** It improves both approaches, and it is the cheapest thing on this page.

---

### Recording protocol

One hold per file, labelled with the existing flag. No new tooling:

```bash
airpod-pose record --out data/poses/triangle_right_tim_03.jsonl \
                   --seconds 15 --label "triangle right"
```

Vary deliberately — a model that knows one body in one session is *worse* than a band, because it will be confidently wrong:

| Vary | Why |
|---|---|
| **People** (≥3) | Necks and ranges of motion differ more than you'd think |
| **Sides** (left/right) | Mirroring gets this cheaply; real data validates the mirror |
| **Re-calibration** | Re-zero between sessions; nothing may depend on one zero |
| **Placement** | Ear vs. forearm are *different datasets*, not one with noise. Label it |
| **Quality** | Deliberate "good" and "collapsed" takes — that's the coaching signal |

**Record the negative class too.** Most of the time nobody is in any pose, and the real job is not confusing "walking to the mat" with Triangle. Budget ~30% of recording time for transitions, standing, fidgeting, adjusting the buds. Bands need this as much as the model does: it is how we find out how often each one false-fires.

### Volume and storage

3 people × 5 poses × 2 sides × 5 reps ≈ **150 holds**, 10 s each ≈ **25 minutes** held. Roughly one focused hour, two people, one room. At ~1.4 MB/minute that's **~35 MB** — fine in git for a hackathon; past ~100 MB, use git-lfs or commit a curated subset.

---

### Representation (where a naive model quietly loses)

**Do not feed Euler angles.** They degenerate near vertical: with 0.5° of real sensor noise, reported roll wobbles **±14° at 88° of pitch** (measured, `docs/sensor-notes.md`). Yoga lives near vertical. Feeding yaw/pitch/roll hands the network pure noise in exactly the poses we care about, and it will fit it.

Feed instead, all **calibration-relative** (`quat.relative(reference, current)`):

1. **Gravity direction in the body frame** — 3 numbers, drift-free, gravity-anchored; what pitch/roll *mean* without the wrap or the singularity.
2. **Relative quaternion** or its rotation-matrix columns — continuous everywhere.
3. **Angular rate magnitude** and rolling std — the steadiness signal; separates "held it" from "passing through it".

**Drop yaw, or augment it away.** It drifts several degrees a minute with no magnetometer anchor. Random yaw rotations as augmentation are better than exclusion: free, and they make the invariance explicit.

**Windowing:** 1.5 s at 50 Hz = 75 samples, stride 0.2 s.

**Augmentation**, by value: random yaw rotation; **left/right mirroring** (doubles the set, gives both sides free, exactly like `mirror_roll` does today); rotation jitter ±3° (simulates calibration error, the most common real perturbation); time warping ±15%; noise matched to *measured* sensor noise, not a guessed sigma.

### Models, cheapest first

| Stage | Model | Params | Role |
|---|---|---|---|
| **0** | Logistic regression / small GBM on window stats | ~100 | The number the CNN must beat |
| **1** | **1D CNN** — 3 conv layers, global pool, softmax | 20–50k | The proposal |
| **2** | GRU / TCN | ~100k | Only if transitions need real sequence modelling |

### Export

Train offline in PyTorch (dev-only, never on the demo path). Export weights to **JSON and run the forward pass in numpy** — ~40 lines for a 3-layer CNN. Keeps the runtime dependency-free (`dependencies = []`) and immune to a broken `pip install` at the venue. Budget: **<2 ms/sample** at 50 Hz, which is enormous for this size.

---

## The bake-off: how we decide

Run both on the *same* recordings, same folds. **Split by person, not by window** — consecutive windows from one hold are near-duplicates, and a random split will report 99% that means nothing. Leave-one-subject-out, reporting per-fold spread rather than just the mean.

| Metric | Why it decides anything |
|---|---|
| Per-class F1 + confusion matrix | Are the mix-ups ones a human would forgive? |
| **Detection latency** | Bands fire in ~20 ms, a window in ~1.5 s. A real regression |
| **False positives per minute in transitions** | Whether a live demo embarrasses us |
| Flicker rate while holding | Hysteresis hides some; measure before it |
| Per-person worst case | A mean of 95% with one person at 60% is a broken demo |

**Decision rule, agreed before we look:** the classifier ships as default only if it beats the bands on per-person worst-case F1 **and** does not regress false-positives-per-minute, with detection latency under 2 s. Otherwise bands stay default and the model stays available behind the flag. If the bands win, that goes in the README as a finding, not a footnote.

### The likely outcome, worth planning for

A **hybrid** is the most plausible winner and costs nothing extra: bands decide `match` (instant, debuggable, no data risk), while a learned model scores *how well* the pose is held — the thing bands genuinely cannot do, and the thing #3 argues is the honest output anyway ("steadier on the left" rather than a verdict). Worth evaluating explicitly rather than treating this as binary.

---

## Integration

- `PoseClassifier` satisfies the same `update()` signature; `error_deg` becomes `(1 − confidence) × scale` so the UI's target arc keeps working unchanged.
- `--matcher bands|model|hybrid`, defaulting to **bands**.
- `HoldTimer` and hysteresis unchanged — orthogonal, already working, and what stops the UI strobing regardless of what decides `match`.

## Risks

| Risk | Mitigation |
|---|---|
| Not enough data → confidently wrong | Stage 0 baseline first; LOSO; bands stay default |
| Overfits one body | ≥3 people; report per-fold spread, not the mean |
| Placement change invalidates everything | Ear and forearm are separate datasets and separate models |
| Latency regresses vs. bands | Measured explicitly; shorter window or a per-sample head |
| Yaw drift leaks in as a feature | Exclude or augment away; verify by ablation |
| Recording eats the hackathon | Timebox to one hour; 150 holds is enough to know |

## Acceptance criteria

- [ ] ≥150 labelled holds, ≥3 people, negative/transition class included, documented naming convention
- [ ] **Bands re-tuned against the same recordings** — the alternative must beat a *measured* baseline, not a guessed one
- [ ] Stage 0 baseline scored leave-one-subject-out
- [ ] 1D CNN trained, exported to JSON, numpy forward pass, <2 ms/sample
- [ ] Head-to-head table: per-class F1, per-person worst case, detection latency, false positives/min, flicker
- [ ] `--matcher` works end to end; default decided by the table, not by preference

## Open questions

1. **Classifying poses, or scoring one pose?** Multi-class needs the full dataset. Scoring one pose is a regression on far less data and may be the better product.
2. **Which placement do we commit to?** Unresolved from #3, and it decides which dataset gets collected. **This should be settled before anyone records.**
3. **Does a bud stream out of the ear?** Still untested, still two minutes with `make monitor`, still gates the forearm option entirely.

---

*Related: #3 (Trikonasana, sensor limits), `docs/sensor-notes.md` (rate, drift, degeneracy numbers), `python/airpod_pose/poses.py` (the incumbent).*
